### PR TITLE
8317515: Unify the code of the parse*() families of methods in j.l.Integer and j.l.Long

### PR DESCRIPTION
**Pull request by rgiulietti**
See the [JBS issue](https://bugs.openjdk.org/browse/JDK-8317515) for the details.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8317515](https://bugs.openjdk.org/browse/JDK-8317515): Unify the code of the parse*() families of methods in j.l.Integer and j.l.Long (**Enhancement** - P4)

### Reviewers
 * [Claes Redestad](https://openjdk.org/census#redestad) (@cl4es - **Reviewer**) ⚠️ Review applies to [bb4cc608](https://git.openjdk.org/jdk/pull/16050/files/bb4cc60873ffa790884e0028db435a415238acac)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16050/head:pull/16050` \
`$ git checkout pull/16050`

Update a local copy of the PR: \
`$ git checkout pull/16050` \
`$ git pull https://git.openjdk.org/jdk.git pull/16050/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16050`

View PR using the GUI difftool: \
`$ git pr show -t 16050`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16050.diff">https://git.openjdk.org/jdk/pull/16050.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16050#issuecomment-1748488675)

### CODE REVIEW

**Comment by wenshao** [comment]
In fact, the most commonly used one is that radix is 10 and coder is latin1, so we can make a fast-path
```java
public final class Integer {
   public static int parseInt(String s) throws NumberFormatException {
        if (s != null && s.coder() == String.LATIN1) {
            byte[] value = s.value();
            int len = value.length;
            if (len == 0) {
                throw new NumberFormatException("");
            }
            int digit = ~0xFF;
            int i = 0;
            byte firstChar = value[i++];
            if (firstChar != '-' && firstChar != '+') {
                digit = DIGITS[firstChar];
            }
            if (digit >= 0 || digit == ~0xFF && len > 1) {
                int limit = firstChar != '-' ? MIN_VALUE + 1 : MIN_VALUE;
                int multmin = limit / 10;
                int result = -(digit & 0xFF);
                boolean inRange = true;
                /* Accumulating negatively avoids surprises near MAX_VALUE */
                while (i < len && (digit = DIGITS[value[i++] & 0xFF]) >= 0
                        && (inRange = result > multmin
                        || result == multmin && digit <= 10 * multmin - limit)) {
                    result = 10 * result - digit;
                }
                if (inRange && i == len && digit >= 0) {
                    return firstChar != '-' ? -result : result;
                }
            }
            throw NumberFormatException.forInputString(s, 10);
        }
        return parseInt(s, 10);
    }

   @Stable
    private static final byte[] DIGITS = new byte[] {
            -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
            -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
            -1, -1, -1, -1, -1, -1, -1, -1,  0,  1,  2,  3,  4,  5,  6,  7,  8,  9, -1, -1,
            -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
            -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
            -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
            -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
            -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
            -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
            -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
            -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
            -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
            -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1 };
}
```

Performance numbers running on MacBook M1 Max are about 9% faster:
```diff
-Benchmark          (size)  Mode  Cnt  Score   Error  Units
-Integers.parseInt     500  avgt   15  2.480 ? 0.017  us/op

+Benchmark          (size)  Mode  Cnt  Score   Error  Units (fast-path)
+Integers.parseInt     500  avgt   15  2.270 ? 0.011  us/op
``

**Comment by rgiulietti** [comment]
The aim of this PR is _not_ to make the code faster, but to make it more uniform. It turns out to be slightly faster even without special cases, but this was not the goal, but just a by-product of the simpler structure.
Over-reliance on micro benchmarks and obsession for speed is not a guidance for this PR.

**Review by cl4es — APPROVED** [review]
Changes look reasonable. The inline logic (such as writing to `inRange` on each loop iteration) makes the logic a bit harder to follow, but I think the improved consistency across implementations trumps this concern. 

I noticed that `parse*(CharSequence..)` and `parseUnsigned` aren't covered by current microbenchmark. That ought to be fixed.

File: src/java.base/share/classes/java/lang/Integer.java [hunk 1]
```diff
@@ -753,31 +734,49 @@ public static int parseUnsignedInt(String s, int radix)
...
+        if (firstChar != '+') {
+            digit = digit(firstChar, radix);
+        }
+        if (digit >= 0 || digit == ~0xFF && len > 1) {
+            int multmax = divideUnsigned(-1, radix);  // -1 is max unsigned int
+            int result = digit & 0xFF;
+            boolean inRange = true;
+            /* Use MIN_VALUE + x < MIN_VALUE + y as unsigned x < y comparison */
+            while (i < len && (digit = digit(s.charAt(i++), radix)) >= 0
+                    && (inRange = MIN_VALUE + result < MIN_VALUE + multmax
```
Comment (merykitty) [inline src/java.base/share/classes/java/lang/Integer.java:767]: `compareUnsigned(result, multmax) < 0` would be better here.

**Comment by merykitty** [comment]
Could we parse the signed int by parsing the unsigned suffix then prepend the sign? It will unify the code path of `parseInt` and `parseUnsignedInt`. 

Thanks.

File: src/java.base/share/classes/java/lang/Integer.java (hunk 1 above)
Comment (rgiulietti) [reply src/java.base/share/classes/java/lang/Integer.java:767]: `MIN_VALUE + result < MIN_VALUE + multmax` is C2-compiled to a single unsigned compare instruction (@merykitty [you should know](https://github.com/openjdk/jdk/pull/6101) ;-) )

**Comment by rgiulietti** [comment]
@merykitty For a `String` input that would mean copying the suffix, which could be quite long, or make use of the method that accepts a `CharSequence`, which has different (although more complete) exception messages. Perhaps in a followup PR.

What would be _really_ nice is to be able to write, for example (leaving apart the exception messages issue above)
```
public static int parseInt(String s, int radix)
                throws NumberFormatException {
    return parseInt(s, 0, s.length(), radix);  // delegate to the CharSequence method
}
```
and leave it to the runtime compiler to perform something similar to
* make a copy of the `parseInt(CharSequence,int,int,int)` code into `parseInt(String,int)`
* adapt the copy to the `String` case by replacing `invokeinterface` for `charAt()` with `invokevirtual`, which can be further optimized to direct invocation because `String` is final, and eventually inlined perfectly.

Currently, that does not seem to happen, but I might be wrong.
This would spare us 4 code duplications here, and perhaps in many other places where we have almost identical methods for `String` and `CharSequence`.

File: src/java.base/share/classes/java/lang/Integer.java (hunk 1 above)
Comment (merykitty) [reply src/java.base/share/classes/java/lang/Integer.java:767]: It's not really about performance, I just think that it is more readable the other way.

**Comment by merykitty** [comment]
@rgiulietti We can do the same thing as the Vector API does, that is to have a private `parseInt(CharSequence, int, int, int)` that is annotated with `@ForceInline`. When called from `parseInt(String, int)`, the compiler, after inlining the former method, can devirtualise all calls to `charAt`, etc. This is like a poor man workaround for monomorphism.

    @ForceInline
    private int parseUnsignedInt(CharSequence, int, int, int) {}

    public int parseUnsignedInt(String s, int radix) {
        return parseUnsignedInt(s, 0, s.length(), radix); // After inlining, this call behaves as if it is a parseUnsignedInt(String, int, int, int)
    }

Thanks.

File: src/java.base/share/classes/java/lang/Integer.java (hunk 1 above)
Comment (rgiulietti) [reply src/java.base/share/classes/java/lang/Integer.java:767]: Well, two lines above the usage there's a comment that should help the reader...

**Comment by rgiulietti** [comment]
@merykitty Interesting. I'll prepare a followup PR in the near future.

**Comment by wenshao** [comment]
The reason parseUnsignedInt(CharSequence s, int beginIndex, int endIndex, int radix) cannot be inlined is because codeSize 350 is greater than default FreqInlineSize 325. Refactoring the exception handling code can be smaller than 325 without using @ForceInline.

**Comment by merykitty** [comment]
@wenshao There are a lot of factors affecting the inlineability of a method and it is not advisable to rely on a method being inlined apart from really simple ones or those that are marked with `@ForceInline`.

**Comment by wenshao** [comment]
Refactoring the code related to exception handling can make the code smaller and easier to be inlined. The code for constructing exceptions in these methods is encapsulated into methods and can be easily maintained together.

https://github.com/rgiulietti/jdk/pull/1

File: src/java.base/share/classes/java/lang/Integer.java [hunk 2]
```diff
@@ -551,52 +551,41 @@ public static int parseInt(String s, int radix)
...
-                }
-
-                if (len == 1) { // Cannot have lone "+" or "-"
-                    throw NumberFormatException.forInputString(s, radix);
-                }
-                i++;
-            }
+        int len = s.length();
+        if (len == 0) {
+            throw new NumberFormatException("");
```
Comment (uschindler) [inline src/java.base/share/classes/java/lang/Integer.java:565]: This caused a regression in Apache Lucene (see https://github.com/apache/lucene/issues/12708) and should be: `throw NumberFormatException.forInputString("", radix);`

File: src/java.base/share/classes/java/lang/Integer.java [hunk 3]
```diff
@@ -632,55 +621,47 @@ public static int parseInt(CharSequence s, int beginIndex, int endIndex, int rad
...
 
-        boolean negative = false;
+        /*
+         * While s can be concurrently modified, it is ensured that each
+         * of its characters is read at most once, from lower to higher indices.
+         * This is obtained by reading them using the pattern s.charAt(i++),
+         * and by not updating i anywhere else.
+         */
+        if (beginIndex == endIndex) {
+            throw new NumberFormatException("");
```
Comment (uschindler) [inline src/java.base/share/classes/java/lang/Integer.java:640]: This caused a regression in Apache Lucene (see https://github.com/apache/lucene/issues/12708) and should be: `throw NumberFormatException.forInputString("", radix);`

File: src/java.base/share/classes/java/lang/Integer.java [hunk 4]
```diff
@@ -812,32 +810,54 @@ public static int parseUnsignedInt(CharSequence s, int beginIndex, int endIndex,
...
+        }
+
+        /*
+         * While s can be concurrently modified, it is ensured that each
+         * of its characters is read at most once, from lower to higher indices.
+         * This is obtained by reading them using the pattern s.charAt(i++),
+         * and by not updating i anywhere else.
+         */
+        if (beginIndex == endIndex) {
             throw new NumberFormatException("");
```
Comment (uschindler) [inline src/java.base/share/classes/java/lang/Integer.java:830]: This caused a regression in Apache Lucene (see https://github.com/apache/lucene/issues/12708) and should be: `throw NumberFormatException.forInputString("", radix);`

File: src/java.base/share/classes/java/lang/Long.java [hunk 5]
```diff
@@ -574,58 +575,47 @@ static int stringSize(long x) {
...
-                }
-
-                if (len == 1) { // Cannot have lone "+" or "-"
-                    throw NumberFormatException.forInputString(s, radix);
-                }
-                i++;
-            }
+        int len = s.length();
+        if (len == 0) {
+            throw new NumberFormatException("");
```
Comment (uschindler) [inline src/java.base/share/classes/java/lang/Long.java:595]: This caused a regression in Apache Lucene (see https://github.com/apache/lucene/issues/12708) and should be: `throw NumberFormatException.forInputString("", radix);`

File: src/java.base/share/classes/java/lang/Long.java [hunk 6]
```diff
@@ -661,55 +651,47 @@ public static long parseLong(CharSequence s, int beginIndex, int endIndex, int r
...
 
-        boolean negative = false;
+        /*
+         * While s can be concurrently modified, it is ensured that each
+         * of its characters is read at most once, from lower to higher indices.
+         * This is obtained by reading them using the pattern s.charAt(i++),
+         * and by not updating i anywhere else.
+         */
+        if (beginIndex == endIndex) {
+            throw new NumberFormatException("");
```
Comment (uschindler) [inline src/java.base/share/classes/java/lang/Long.java:670]: This caused a regression in Apache Lucene (see https://github.com/apache/lucene/issues/12708) and should be: `throw NumberFormatException.forInputString("", radix);`

**Review by uschindler — COMMENTED** [review]
This caused issue sin Apache Lucene, as it should not chnge message as it was befor ethis change.

See https://bugs.openjdk.org/browse/JDK-8318646

**Comment by rgiulietti** [comment]
@uschindler The spec of `parse[Int|Long]` is committed to throw particular exception types, but says nothing about the message content.
IMO, relying on exception messages should be considered as a last resort choice, as they can change.
 
However, I agree that this is an easy fix.

**Comment by uschindler** [comment]
> @uschindler The spec of `parse[Int|Long]` is committed to throw particular exception types, but says nothing about the message content. IMO, relying on exception messages should be considered as a last resort choice, as they can change.
> 
> However, I agree that this is an easy fix.

Yes. I agree with this. 

The issue why I raised thus was that there may be code that relies on this. Although it is not part of spec, it might be a good idea to also keep code consistent and use a single way to instantiate the NumberFormatException.

**Pull request closed without merging**